### PR TITLE
PKCS#11 Ed25519 tests

### PR DESCRIPTION
**Pull request by varder**
Adds tests to import, sign and verify Ed25519, Ed25519ctx and Ed25519ph.

The pull request comes with the following pull requests:
https://github.com/OP-TEE/optee_os/pull/5559
https://github.com/OP-TEE/optee_os/pull/5574

### CODE REVIEW

File: host/xtest/pkcs11_1000.c [hunk 1]
```diff
@@ -7797,3 +7797,189 @@ static void xtest_pkcs11_test_1024(ADBG_Case_t *c)
...
+struct eddsa_test {
+	CK_BYTE *ec_params;
+	size_t ec_params_len;
+	const uint8_t *message;
+	size_t message_len;
+	const uint8_t *private;
+	size_t private_len;
+	const uint8_t *public;
+	size_t public_len;
+	const bool flag;
```
Comment (etienne-lms) [inline host/xtest/pkcs11_1000.c:7836]: rename `ph_flag` as this is the pre-hash flag?

File: host/xtest/pkcs11_1000.c [hunk 2]
```diff
@@ -7797,3 +7797,189 @@ static void xtest_pkcs11_test_1024(ADBG_Case_t *c)
...
+
+	CK_MECHANISM sign_mechanism = {
+		CKM_EDDSA,
+		&eddsa_params,
+		sizeof(eddsa_params),
+	};
+
+	CK_ATTRIBUTE public_key_template[] = {
+		{ CKA_CLASS, &(CK_OBJECT_CLASS){ CKO_PUBLIC_KEY }, sizeof(CK_OBJECT_CLASS) },
+		{ CKA_KEY_TYPE,	&(CK_KEY_TYPE){CKK_EC_EDWARDS}, sizeof(CK_KEY_TYPE) },
```
Comment (etienne-lms) [inline host/xtest/pkcs11_1000.c:7881]: nit: space char around `CKK_EC_EDWARDS`: `{ CKK_EC_EDWARDS }`

File: host/xtest/pkcs11_1000.c [hunk 3]
```diff
@@ -7797,3 +7797,189 @@ static void xtest_pkcs11_test_1024(ADBG_Case_t *c)
...
+};
+
+static CK_BYTE ed25519_params[] = {
+	0x06, 0x09, 0x2b, 0x06, 0x01, 0x04, 0x01, 0xda,
+	0x47, 0x0f, 0x01,
+};
+
+static struct eddsa_test eddsa_sign_tests[] = {
+	CKTEST_EDDSA_TEST(ed25519_params, ed25519_rfc_8032_7_1, 0),
+	CKTEST_EDDSA_TEST(ed25519_params, ed25519_rfc_8032_7_1, 1),
```
[the remark this answers is not in the document]
Comment (varder) [reply host/xtest/pkcs11_1000.c]: agree, removed

File: host/xtest/pkcs11_1000.c (hunk 2 above)
Comment (varder) [reply host/xtest/pkcs11_1000.c]: fixed

File: host/xtest/pkcs11_1000.c [hunk 4]
```diff
@@ -7797,3 +7797,189 @@ static void xtest_pkcs11_test_1024(ADBG_Case_t *c)
...
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto err;
+		}
+
+		rv = C_SignInit(session, &sign_mechanism, private_key);
+		if (!ADBG_EXPECT_CK_OK(c, rv))
+			goto err;
+
+		rv = C_Sign(session, (CK_BYTE_PTR)test->message, test->message_len,
+					 (CK_BYTE_PTR)sign, &sign_len);
```
[the remark this answers is not in the document]
Comment (varder) [reply host/xtest/pkcs11_1000.c]: fixed

File: host/xtest/pkcs11_1000.c [hunk 5]
```diff
@@ -7797,3 +7797,189 @@ static void xtest_pkcs11_test_1024(ADBG_Case_t *c)
...
+        .ec_params_len = ARRAY_SIZE(_ec_params), \
+        .message = _vect ##_ptx, \
+        .message_len = ARRAY_SIZE(_vect ##_ptx), \
+        .private = _vect ##_private, \
+        .private_len = ARRAY_SIZE(_vect ##_private), \
+        .public = _vect ##_public, \
+        .public_len = ARRAY_SIZE(_vect ##_public), \
+		.flag = _flag, \
+		.context = _vect ## _context, \
+		.context_len = ARRAY_SIZE(_vect ##_context), \
```
[the remark this answers is not in the document]
Comment (varder) [reply host/xtest/pkcs11_1000.c]: fixed

File: host/xtest/pkcs11_1000.c (hunk 1 above)
Comment (varder) [reply host/xtest/pkcs11_1000.c]: agree

**Comment by varder** [comment]
Hi @etienne-lms,
Could you please have a look at the updates?
Thanks

File: host/xtest/pkcs11_1000.c [hunk 6]
```diff
@@ -7797,3 +7797,211 @@ static void xtest_pkcs11_test_1024(ADBG_Case_t *c)
...
+					       private_key_template,
+					       ARRAY_SIZE(private_key_template),
+					       &public_key, &private_key);
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto err;
+		}
+
+		rv = C_SignInit(session, &sign_mechanism, private_key);
+		if (!ADBG_EXPECT_CK_OK(c, rv))
+			goto err;
```
Comment (etienne-lms) [inline host/xtest/pkcs11_1000.c:7979]: failing here needs to attempt to destroy created persistent object(s).

File: host/xtest/pkcs11_1000.c [hunk 7]
```diff
@@ -7797,3 +7797,211 @@ static void xtest_pkcs11_test_1024(ADBG_Case_t *c)
...
+
+	rv = init_lib_and_find_token_slot(&slot);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		return;
+
+	rv = C_OpenSession(slot,
+			   CKF_SERIAL_SESSION | CKF_RW_SESSION,
+			   NULL, NULL, &session);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto err;
```
Comment (etienne-lms) [inline host/xtest/pkcs11_1000.c:7925]: `goto err_close_lib;`
and only call `close_lib()`

**Review by etienne-lms — COMMENTED** [review]
Sorry @varder for the late feedback. Error path issues to address.

File: host/xtest/pkcs11_1000.c (hunk 7 above)
Comment (varder) [reply host/xtest/pkcs11_1000.c]: done

**Comment by varder** [comment]
Hi @etienne-lms 
Can I squash the commit?

**Review by etienne-lms — COMMENTED** [review]
> Can I squash the commit?

Ok.
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Pull request merged**